### PR TITLE
CBE: parameters

### PR DESCRIPTION
**Pull request by pixelherodev**

### CODE REVIEW

**Comment by pixelherodev** [comment]
This is okay as-is for the given tests, but I noticed a serious design issue that needs to be addressed before I do any more feature work.

The IntCast has its own ZIR, references the arg, and is referenced by the inline asm. *The intcast is referenced **twice***. First, as its own Inst, then as an operand of the inline asm. For intCasts, whose results are ignored in a statement (and thus, one reference can be safely ignored), it's fine as is, but for more complex ZIR, the CBE as a whole is likely going to require some rethinking.

File: src-self-hosted/codegen/c.zig [hunk 1]
```diff
@@ -41,16 +40,33 @@ fn renderType(file: *C, writer: std.ArrayList(u8).Writer, T: Type, src: usize) !
     }
 }
 
+fn renderValue(file: *C, writer: std.ArrayList(u8).Writer, val: Value, src: usize) !void {
+    switch (val.tag()) {
```
Comment (andrewrk) [inline src-self-hosted/codegen/c.zig:44]: It's OK to switch on the tag for a "hot path" but really, this should be switching on the `zigTypeTag()` and dealing with the value on a per-type basis. When you write skeleton code like this, it's good to keep in mind the contributors that will look at this code after you and try to improve it. As is, the structure here is misleading to contributors. I suggest to switch on the type, and check if it's an unsigned integer 64 bits or less, and then call `toUnsignedInt`.

File: src-self-hosted/codegen/c.zig [hunk 2]
```diff
@@ -88,15 +104,21 @@ fn genFn(file: *C, decl: *Decl) !void {
...
-                .assembly => try genAsm(file, inst.castTag(.assembly).?, decl),
+                .assembly => try genAsm(file, inst.castTag(.assembly).?, decl, &argdex),
                 .call => try genCall(file, inst.castTag(.call).?, decl),
                 .ret => try genRet(file, inst.castTag(.ret).?, decl, tv.ty.fnReturnType()),
                 .retvoid => try file.main.writer().print("    return;\n", .{}),
+                .arg => {},
                 .dbg_stmt => try genDbgStmt(file, inst.castTag(.dbg_stmt).?, decl),
+                .breakpoint => try genBreak(file, inst.castTag(.breakpoint).?, decl),
+                .unreach => try genUnreach(file, inst.castTag(.unreach).?, decl),
+                // This will be handled correctly later?
```
Comment (andrewrk) [inline src-self-hosted/codegen/c.zig:120]: what's this talking about?

Comment (pixelherodev) [reply src-self-hosted/codegen/c.zig]: Nothing now :)

**Comment by pixelherodev** [comment]
Looked at codegen.zig; I was right about needing an Inst -> name mapping. New code:

```c
zig_noreturn void exit(uint8_t arg0) {
    const size_t __temp_0 = (size_t)arg0;
    register size_t rax_constant __asm__("rax") = 231;
    register size_t rdi_constant __asm__("rdi") = __temp_0;
    __asm volatile ("syscall" :: ""(rax_constant), ""(rdi_constant));
    zig_unreachable();
}
```

This fully addresses the issue I was talking about; args, int casts, and all other IR dependencies should be 100% good now.

**Comment by pixelherodev** [comment]
Still have a bit more to do on this before it's ready for review again. Re-marking as draft.

File: src-self-hosted/link.zig [hunk 3]
```diff
@@ -229,7 +228,7 @@ pub const File = struct {
             return &c_file.base;
         }
 
-        pub fn fail(self: *C, src: usize, comptime format: []const u8, args: anytype) !void {
+        pub fn fail(self: *C, src: usize, comptime format: []const u8, args: anytype) anyerror {
```
Comment (andrewrk) [inline src-self-hosted/link.zig:231]:
```suggestion
        pub fn fail(self: *C, src: usize, comptime format: []const u8, args: anytype) error{AnalysisFail, OutOfMemory} {
```

File: src-self-hosted/codegen/c.zig [hunk 4]
```diff
@@ -91,27 +138,60 @@ fn genFn(file: *C, decl: *Decl) !void {
...
+}
+
+fn genIntCast(ctx: *Context, inst: *Inst.UnOp) !?[]u8 {
+    if (inst.base.isUnused())
+        return null;
+    const op = inst.operand;
+    const writer = ctx.file.main.writer();
+    const name = try ctx.name();
+    const from = ctx.inst_map.get(op) orelse
+        return ctx.file.fail(ctx.decl.src(), "Internal error in C backend: intCast argument not found in inst_map", .{});
```
Comment (andrewrk) [inline src-self-hosted/codegen/c.zig:183]: You're for sure going to want to make this a function and then call it from pretty much every single IR instruction's gen function. *That* is where you will handle constants - with possibly some specific overrides for specific situations.

I'm ok with merging this as is for now though.

**Pull request closed without merging**